buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,316 @@
+                directories.poll()
+                traverse(directories, f, functionMap)
+            }
+        }
+    }
+
+    /**
+     * Extracts the functions from the given file into functionMap. Parses
+     * functions of the form:
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     * @param file The file to process
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun processFile(file: File, functionMap: MutableMap<String, String>) {
+        var fileText = file.readText(Charsets.UTF_8)
+        val declarations: Queue<MatchResult> = LinkedList(declarationPattern.findAll(fileText).toList())
+        val tokens: Queue<String> = LinkedList(tokenizer.split(fileText))
+        var charIndex = 0
+        while (declarations.isNotEmpty()) {
+            val declaration = declarations.poll()
+            while (charIndex < declaration.range.start) {
+                charIndex += tokens.poll().length
+            }
+            charIndex = processFunction(tokens, functionMap, charIndex,
+                declaration.range.endInclusive, getFunctionName(declaration.value))
+        }
+    }
+
+    /**
+     * Parses a function of the form:
+     *
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     *
+     * @param tokens The tokens starting from the function declaration and
+     *               ending at EOF
+     * @param functionMap The map to which the extracted functions will be added to
+     * @param charIndex The character index of the entire text of the file at which
+     *                  the parsing is commencing at
+     * @param declarationEnd The char index (of entire file text) at which the l
+     *                       function declaration ends
+     * @param functionName The name of the function being processed
+     */
+    private fun processFunction(
+        tokens: Queue<String>,
+        functionMap: MutableMap<String, String>,
+        charIndex: Int,
+        declarationEnd: Int,
+        functionName: String
+    ): Int {
+        var updatedCharIndex = charIndex
+        val functionBuilder = StringBuilder()
+        // Process the declaration:
+        while (updatedCharIndex < declarationEnd && tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
+            updatedCharIndex += token.length
+        }
+        // Called twice as there are two curly blocks.
+        updatedCharIndex = processCurlyBlock(functionBuilder, tokens, updatedCharIndex)
+        updatedCharIndex = processCurlyBlock(functionBuilder, tokens, updatedCharIndex)
+
+        functionMap.put(functionName, functionBuilder.toString())
+        return updatedCharIndex
+    }
+
+    /**
+     * Parses a block of text surrounded by curly braces. The function keeps track
```

Comment:
We should say that this function assumes the function definition/content are in a valid format.